components/content/content-editor/controls/mention/autocomplete/controls.ts

```diff
@@ -0,0 +1,254 @@
+					this.visible = true;
+
+					return;
+				}
+			}
+		}
+		this.visible = false;
+	}
+
+	private async updateSuggestions(suggestion: string) {
+		this.users = MentionCache.getUsers(this.$route, suggestion).sort((a, b) =>
+			numberSort(b.match, a.match)
+		);
+
+		this.remoteSuggestions(suggestion);
+
+		if (this.inverted) {
+			this.selectedIndex = this.users.length - 1;
+
+			// If we are inverted, wait for the list to render and then scroll to the bottom.
+			await this.$nextTick();
+			if (this.$refs.list) {
+				this.$refs.list.scrollTop = this.$refs.list.scrollHeight;
+			}
+		} else {
+			this.selectedIndex = 0;
+		}
+	}
+
+	private remoteSuggestions(suggestion: string) {
+		// Stop any existing queued search timeout.
+		if (this.remoteSuggestionTimeout) {
+			clearTimeout(this.remoteSuggestionTimeout);
+			this.remoteSuggestionTimeout = undefined;
+			this.isLoading = false;
+		}
+
+		// If we already have the search results cached, use those and don't query backend.
+		if (this.searchResultsCache[suggestion]) {
+			const searchUsers = this.searchResultsCache[suggestion];
+			this.processSearchUsers(searchUsers);
+		} else {
+			// We set a timeout for 200ms here to not send right away when the user is typing fast.
+			this.remoteSuggestionTimeout = setTimeout(async () => {
+				this.isLoading = true;
+
+				const payload = await Api.sendRequest(
+					'/web/search/mention-suggestions?q=' + encodeURIComponent(suggestion),
+					undefined,
+					{ detach: true }
+				);
+
+				if (payload.users) {
+					const searchUsers = User.populate(payload.users);
+
+					// Add to cache
+					if (!this.searchResultsCache[suggestion]) {
+						this.searchResultsCache[suggestion] = searchUsers;
+					}
+
+					// Only process results if the currently active suggestion is still the one we initiated the search with.
+					if (suggestion === this.suggestion) {
+						this.processSearchUsers(searchUsers);
+					}
+				}
+
+				this.isLoading = false;
+			}, 200);
+		}
+	}
+
+	private processSearchUsers(searchUsers: User[]) {
+		let newUserCount = 0;
+		for (const searchUser of searchUsers) {
+			const existingUser = this.users.find(i => i.user.id === searchUser.id);
+			// If the user already exists in the results, increase their match ranking.
+			if (existingUser) {
+				existingUser.match++;
+			} else {
+				const newUser = { user: searchUser, rank: 0, match: 0, source: 'search' };
+				newUser.match = MentionCache.calculateUserMatch(this.suggestion, newUser);
+				this.users.push(newUser);
+				// Push down the selected index in inverted since we append at the end.
+				if (this.inverted) {
+					newUserCount++;
+				}
+			}
+		}
+		// Sort after inserting
+		this.users = this.users.sort((a, b) => numberSort(b.match, a.match));
+
+		this.selectedIndex += newUserCount;
+	}
+
+	isSelected(userId: number) {
+		return this.users[this.selectedIndex].user.id === userId;
+	}
+
+	async onKeyDown(e: KeyboardEvent) {
```

Comment:
Think it'd be nice to make the selection loop around when it gets to the border instead of limiting it?